### PR TITLE
Initial support for RHEL8 as a dev environment.

### PR DESCRIPTION
**Pull request by russellb**
This PR and https://github.com/metal3-io/metal3-dev-env/pull/19 contain the changes I needed to make to get dev-scripts to run on top of RHEL8.  The types of changes include:

* A number of package names are different, mostly Python related.

* A number of packages aren't available.  There's no EPEL for RHEL8 and there's also not an OSP version we can subscribe to.  This isn't a big deal.  We install a few things from source until we can pull them from OSP, and none of these things are ones that would be needed on a production environment anyway.

* The firewall setup didn't work for me.  I found it easiest to fix it using firewalld.  I suspect this has to do with changes in behavior since RHEL8 uses nftables instead of iptables.  firewalld seems like a much nicer way to manage these rules anyway, even for the CentOS 7 environment, so I suggest we move everything to firewalld over time.

I have one remaining quirk that i haven't fixed.  I have to manually kill some `vbmc` processes after running `make clean` before running `make` again.  Otherwise, the cluster does come up successfully.

```sh
$ oc get nodes
NAME       STATUS   ROLES    AGE   VERSION
master-0   Ready    master   30m   v1.13.4+cb455d664
master-1   Ready    master   30m   v1.13.4+cb455d664
master-2   Ready    master   30m   v1.13.4+cb455d664
worker-0   Ready    worker   63s   v1.13.4+cb455d664
```

### CODE REVIEW

**Comment by derekhiggins** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/721/

File: 01_install_requirements.sh [hunk 1]
```diff
@@ -24,65 +35,133 @@ fi
...
-# We're reusing some tripleo pieces for this setup so clone them here
-cd
-if [ ! -d tripleo-repos ]; then
-  git clone https://git.openstack.org/openstack/tripleo-repos
-fi
-pushd tripleo-repos
-sudo python setup.py install
-popd
+if [ "${RHEL8}" = "y" ] ; then
+    sudo subscription-manager repos --enable=ansible-2-for-rhel-8-x86_64-rpms
```
Comment (mcornea) [inline 01_install_requirements.sh]: Can we make this optional or check if the node is subscribed to allow it working on non-subscribed nodes(assuming they're using manually set repos).

File: 01_install_requirements.sh [hunk 2]
```diff
@@ -13,9 +13,20 @@ fi
...
+
 # Install EPEL required by some packages
 if [ ! -f /etc/yum.repos.d/epel.repo ] ; then
-    if grep -q "Red Hat Enterprise Linux" /etc/redhat-release ; then
+    if [ "${RHEL8}" = "y" ] ; then
+        # TODO(russellb) Fix this when EPEL 8 is available
+        # sudo dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
+        #
+        # It's also possible we never need EPEL and everything we would have pulled from there is
+        # available through OSP, instead.  There's no OSP release for RHEL 8 yet either, though.
```
Comment (dtantsur) [inline 01_install_requirements.sh]: I think RDO Train will support CentOS 8 soon(ish)

File: 01_install_requirements.sh [hunk 3]
```diff
@@ -13,9 +13,20 @@ fi
...
+fi
+
 # Install EPEL required by some packages
 if [ ! -f /etc/yum.repos.d/epel.repo ] ; then
-    if grep -q "Red Hat Enterprise Linux" /etc/redhat-release ; then
+    if [ "${RHEL8}" = "y" ] ; then
+        # TODO(russellb) Fix this when EPEL 8 is available
+        # sudo dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
+        #
+        # It's also possible we never need EPEL and everything we would have pulled from there is
```
Comment (dtantsur) [inline 01_install_requirements.sh]: We better not use EPEL in our final product... We probably need it because of `pip` and we shouldn't really be using `pip` ;)

File: 01_install_requirements.sh (hunk 1 above)
Comment (dtantsur) [reply 01_install_requirements.sh]: ++

File: 01_install_requirements.sh [hunk 4]
```diff
@@ -24,65 +35,133 @@ fi
...
+      libguestfs-tools \
+      libvirt \
+      libvirt-devel \
+      libvirt-daemon-kvm \
+      podman \
+      qemu-kvm \
+      virt-install \
+      unzip \
+      network-scripts \
+      ipmitool
```
Comment (dtantsur) [inline 01_install_requirements.sh]: Why is this needed?

File: 01_install_requirements.sh [hunk 5]
```diff
@@ -24,65 +35,133 @@ fi
...
-sudo curl -sL https://dl.yarnpkg.com/rpm/yarn.repo -o /etc/yum.repos.d/yarn.repo
-curl -sL https://rpm.nodesource.com/setup_10.x | sudo bash -
+    # TODO(russellb) - Install an rpm for this once OSP for RHEL8 is out
+    pushd ~
+    if [ ! -d virtualbmc ] ; then
+        git clone https://git.openstack.org/openstack/virtualbmc
+    fi
+    pushd virtualbmc
+    git pull -r
+    sudo pip3 install .
```
Comment (dtantsur) [inline 01_install_requirements.sh]: This is certainly going to pull in a lot of packages.. If `--user` doesn't work for some reason, let's use a virtual environment.

File: 01_install_requirements.sh [hunk 6]
```diff
@@ -24,65 +35,133 @@ fi
...
+    popd ; popd
+else
+    # We're reusing some tripleo pieces for this setup so clone them here
+    cd
+    if [ ! -d tripleo-repos ]; then
+      git clone https://git.openstack.org/openstack/tripleo-repos
+    fi
+    pushd tripleo-repos
+    sudo python setup.py install
```
Comment (dtantsur) [inline 01_install_requirements.sh]: I don't think tripleo-repos requires to be installed, it's one script

File: 01_install_requirements.sh [hunk 7]
```diff
@@ -24,65 +35,133 @@ fi
...
+
+    # Needed to get a recent python-virtualbmc package
+    sudo tripleo-repos current-tripleo
+
+    # There are some packages which are newer in the tripleo repos
+    sudo yum -y update
+
+    # Setup yarn and nodejs repositories
+    sudo curl -sL https://dl.yarnpkg.com/rpm/yarn.repo -o /etc/yum.repos.d/yarn.repo
+    curl -sL https://rpm.nodesource.com/setup_10.x | sudo bash -
```
Comment (dtantsur) [inline 01_install_requirements.sh]: Sigh...

File: 01_install_requirements.sh [hunk 8]
```diff
@@ -24,65 +35,133 @@ fi
...
+    # TODO(russellb) - Install an rpm for this once OSP for RHEL8 is out
+    sudo dnf groupinstall -y "Development Tools"
+    sudo dnf install -y python36-devel
+    pushd ~
+    if [ ! -d openstackclient ] ; then
+        git clone https://git.openstack.org/openstack/openstackclient
+    fi
+    pushd openstackclient
+    git pull -r
+    sudo pip3 install .
```
Comment (dtantsur) [inline 01_install_requirements.sh]: This is going to pull in a lot of packages, randomly overriding system ones. Even for a temporary hack, I think we should use `--user` or a virtual env. By the way, if you want to have ironic commands, you have to install `ironicclient` and `ironic-inspector-client` as well.

File: 02_configure_host.sh [hunk 9]
```diff
@@ -121,23 +125,44 @@ fi
...
+        sudo firewall-cmd --zone=libvirt --add-port=${port}/udp --permanent
+    else
+        if ! sudo iptables -C INPUT -i provisioning -p udp --dport $port -j ACCEPT 2>/dev/null ; then
+            sudo iptables -I INPUT -i provisioning -p udp --dport $port -j ACCEPT
+        fi
     fi
 done
 
+# mDNS
+if [ "${RHEL8}" = "y" ] ; then
```
Comment (dtantsur) [inline 02_configure_host.sh]: In theory, we should be doing it for RHEL7 as well. However, I suspect our firewall is wide open by default, and all these `iptables` lines are not needed..

**Comment by schmaustech** [comment]
Even with this patch applied I seem to hit the following issue:

+ sudo yum -y update
Updating Subscription Management repositories.
Last metadata expiration check: 0:21:58 ago on Fri 14 Jun 2019 07:16:36 AM CDT.
Error: 
 Problem: cannot install the best update candidate for package libidn2-2.0.5-1.el8.x86_64
  - nothing provides libunistring.so.0()(64bit) needed by libidn2-2.2.0-1.el7.x86_64
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

**Comment by schmaustech** [comment]
This patch no longer appears to apply cleanly:
``[bschmaus@rhel8-ocp-boot dev-scripts]$ /usr/bin/curl -L https://github.com/openshift-metal3/dev-scripts/pull/610.patch | git am
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   154    0   154    0     0    696      0 --:--:-- --:--:-- --:--:--   696
100 16835    0 16835    0     0  28485      0 --:--:-- --:--:-- --:--:-- 28485
Applying: rhel8: First pass at installing requirements.
error: patch failed: 01_install_requirements.sh:33
error: 01_install_requirements.sh: patch does not apply
Patch failed at 0001 rhel8: First pass at installing requirements.
Use 'git am --show-current-patch' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
``

**Comment by stbenjam** [comment]
@schmaustech I've rebased this PR, so it should apply cleanly for you

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/768/

File: 01_install_requirements.sh [hunk 10]
```diff
@@ -13,9 +13,20 @@ fi
 # Update to latest packages first
 sudo yum -y update
 
+if grep -q "Red Hat Enterprise Linux release 8" /etc/redhat-release 2>/dev/null ; then
```
[the remark this answers is not in the document]
Comment (russellb) [reply 01_install_requirements.sh]: Can this be a follow-up?

File: 01_install_requirements.sh [hunk 11]
```diff
@@ -13,9 +13,20 @@ fi
...
+    RHEL8=y
```
[the remark this answers is not in the document]
Comment (russellb) [reply 01_install_requirements.sh]: I've changed this to "True" and quoted it.

File: 01_install_requirements.sh (hunk 3 above)
Comment (russellb) [reply 01_install_requirements.sh]: Of course not.

I think it's only added in the centos 7 case because we pull some stuff from RDO, and RDO assumes EPEL is present, as well for some things.

File: 01_install_requirements.sh (hunk 2 above)
Comment (russellb) [reply 01_install_requirements.sh]: sounds good.

File: 01_install_requirements.sh (hunk 1 above)
Comment (russellb) [reply 01_install_requirements.sh]: If someone wants to do that as a follow-up, that's fine.  My intention is to test exactly as a customer would, using an official release, and registered/subscribed.  It works fine with my employee account.

File: 01_install_requirements.sh (hunk 4 above)
Comment (russellb) [reply 01_install_requirements.sh]: I don't actually remember.  I think I added it just for testing purposes as I was debugging things, and it seemed helpful to have around.

File: 01_install_requirements.sh (hunk 8 above)
Comment (russellb) [reply 01_install_requirements.sh]: The ironic libraries seemed to have been installed automatically.  I have them on my dev box, anyway, and have only been installing things via these scripts.

File: 01_install_requirements.sh (hunk 5 above)
Comment (russellb) [reply 01_install_requirements.sh]: meh, I'm not that bothered by the global pip install stuff personally.  I'm fine if someone wants to clean this up as a follow-up.

We'll switch this to an rpm as soon as one is available.  We could probably just build one ourselves for this one.

File: 01_install_requirements.sh (hunk 6 above)
Comment (russellb) [reply 01_install_requirements.sh]: This is pre-existing code that got moved

File: 01_install_requirements.sh (hunk 7 above)
Comment (russellb) [reply 01_install_requirements.sh]: This is pre-existing code that got moved

Comment (russellb) [reply 01_install_requirements.sh]: ... and can probably just be removed in a follow-up.  It was added for facet, which doesn't get installed and is on ice anyway.

File: 02_configure_host.sh (hunk 9 above)
Comment (russellb) [reply 02_configure_host.sh]: Yeah, I just didn't want to break centos7 envs just yet.  I figured I'd get it working on rhel8 with firewall-cmd and then come back later and test using that on centos7.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/794/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/807/

**Comment by schmaustech** [comment]
When this patch is applied now for RHEL8 it makes the following changes:

it changes this line(s) from:

if [ "$MANAGE_PRO_BRIDGE" == "y" ]; then

to 

if [ "$MANAGE_PRO_BRIDGE" == "True" ]; then

and

if [ "$MANAGE_INT_BRIDGE" == "y" ]; then

to

if [ "$MANAGE_INT_BRIDGE" == "True" ]; then

This breaks the logic and the provisioning and baremetal interfaces never get defined.

Can we get this cleaned up and I can run another test to validate?

**Comment by russellb** [comment]
@schmaustech sorry about that, should be good now

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/809/

**Comment by hardys** [comment]
Ok this passed CI and lgtm for a first-pass, provided @schmaustech can confirm all works OK I'd suggest we merge this and https://github.com/metal3-io/metal3-dev-env/pull/19 then address any remaining issues as follow-ups?

**Comment by schmaustech** [comment]
I can confirm it works as well.
